src/python/pants/backend/docker/util_rules/docker_build_context.py

```diff
@@ -332,7 +339,52 @@ async def create_docker_build_context(
     else:
         logger.debug("Did not build any packages for Docker image")
 
-    embedded_pkgs_digest = [built_package.digest for built_package in embedded_pkgs]
+    # Create digests for embedded packages. For upstream Docker images, we use only the image ID
+    # to ensure hash stability. This prevents changes in image tags (which may include timestamps
+    # or other dynamic values) from affecting the build context hash of dependent images.
+    embedded_pkgs_digest = []
+
+    # For Docker images, we need to extract the metadata filename and create stable digests
+    docker_metadata_gets = []
+    docker_package_indices = []
+    for i, built_package in enumerate(embedded_pkgs):
```

Comment:
How about `for field_set, built_package in zip(pkgs_wanting_embedding, embedded_pkgs)`, and maintain a list of `docker_packages` instead of `docker_package_indices`?  The index is one extra thing to keep track of for the reader, and we could get rid of it.